### DISCUSSION TITLE
Help with using infinispan

### DISCUSSION THREAD

**Discussion by eldarj** [infinispan/infinispan#10224] [Q&A] [answered]
I'm looking into the docs but I can't come up with an easy way to retrieve and store simple data within a cluster of apps running Infinispan.

I'd like to store a simple String and key-value pair data - and retrieve the same, and ideally there wouldn't be any NIO (remote gets) if the _get_ is running on the same node where the cache entry is stored.

I'm not sure how replication runs on Infinispan ie. does it store data replicas on all nodes or just some nodes.

I'm particularly using Keycloak in HA, which internally runs a cluster of infinispan nodes.

Can this be achieved out of the box?

**Comment by tristantarrant** [comment]
"Replicated" caches are exactly what you want: they store a copy of the data on each node. Reads are local. 
Read https://infinispan.org/docs/stable/titles/configuring/configuring.html#cache-modes_caches for info about the different caching modes.

As far as actually interacting with a cache, look at our embedded tutorial: https://infinispan.org/tutorials/embedded/index.html

**Comment by eldarj** [comment]
@tristantarrant Thanks for the tips, I'll check out the embedded tutorial. Regarding the cache mode, I'll have to check what mode is configured on Keycloak though.

**Comment by tristantarrant** [comment, marked as answer]
The mode is per-cache
